cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala

```diff
@@ -24,21 +24,21 @@ import scala.util.Random
 /**
  * A Traverser traverses the graph in a certain order of nodes.
  */
-trait Traverser extends Iterator[Node] { self =>
+trait Traverser[+V <: Node] extends Iterator[V] { self =>
   /**
    * Traverser only visits node Ids listed in some node's edges, and thus
    * we know that node must exist, safe to assume graph.getNodeById return non-None value
    */
-  protected def getExistingNodeById(graph: Graph, id: Int) = graph.getNodeById(id).get
+  protected def getExistingNodeById[V <: Node](graph: Graph[V], id: Int) = graph.getNodeById(id).get
```

Comment:
No I think W >: V is right. A traverser of type V can do something with
graph of type V or supertype but can not of subtypes.

On Saturday, February 7, 2015, Anish Shah <notifications@github.com> wrote:

> In
> cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala
> <https://github.com/twitter/cassovary/pull/148#discussion_r24290377>:
>
> >    /**
> >     * Traverser only visits node Ids listed in some node's edges, and thus
> >     * we know that node must exist, safe to assume graph.getNodeById return non-None value
> >     */
> > -  protected def getExistingNodeById(graph: Graph, id: Int) = graph.getNodeById(id).get
> > +  protected def getExistingNodeById[V <: Node](graph: Graph[V], id: Int) = graph.getNodeById(id).get
>
> It should be W <: Node, I guess.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/cassovary/pull/148/files#r24290377>.
>